testsuite/model/src/test/java/org/keycloak/testsuite/model/transaction/StorageTransactionTest.java

```diff
@@ -0,0 +1,203 @@
+
+    @Override
+    protected void createEnvironment(KeycloakSession s) {
+        RealmModel oldRealm = s.realms().getRealmByName("1");
+        if (oldRealm != null) {
+            s.realms().removeRealm(oldRealm.getId());
+        }
+
+        RealmModel r = s.realms().createRealm("1");
+        r.setDefaultRole(s.roles().addRealmRole(r, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + r.getName()));
+        r.setAttribute("k1", "v1");
+        realmId = r.getId();
+    }
+
+    @Override
+    protected void cleanEnvironment(KeycloakSession s) {
+        s.realms().removeRealm(realmId);
+    }
+
+    @Override
+    protected boolean isUseSameKeycloakSessionFactoryForAllThreads() {
+        return true;
+    }
+
+    @Test
+    public void testTwoTransactionsSequentially() throws Exception {
+        TransactionController tx1 = new TransactionController(getFactory());
+        TransactionController tx2 = new TransactionController(getFactory());
+
+        tx1.begin();
+        assertThat(
+                tx1.runStep(session -> {
+                    session.realms().getRealm(realmId).setAttribute("k2", "v1");
+                    return session.realms().getRealm(realmId).getAttribute("k2");
+                }), equalTo("v1"));
+        tx1.commit();
+
+        tx2.begin();
+        assertThat(
+                tx2.runStep(session -> session.realms().getRealm(realmId).getAttribute("k2")),
+                equalTo("v1"));
+        tx2.commit();
+
+    }
+
+    @Test
+    public void testRepeatableRead() {
+        TransactionController tx1 = new TransactionController(getFactory());
+        TransactionController tx2 = new TransactionController(getFactory());
+        TransactionController tx3 = new TransactionController(getFactory());
+
+        tx1.begin();
+        tx2.begin();
+        tx3.begin();
+
+        // Read original value in tx1
+        assertThat(
+                tx1.runStep(session -> session.realms().getRealm(realmId).getAttribute("k1")),
+                equalTo("v1"));
+
+        // change value to new in tx2
+        tx2.runStep(session -> {
+            session.realms().getRealm(realmId).setAttribute("k1", "v2");
+            return null;
+        });
+        tx2.commit();
+
+        // tx1 should still return the value that already read
+        assertThat(
+                tx1.runStep(session -> session.realms().getRealm(realmId).getAttribute("k1")),
+                equalTo("v1"));
+
+        // tx3 should return the new value
+        assertThat(
+                tx3.runStep(session -> session.realms().getRealm(realmId).getAttribute("k1")),
+                equalTo("v2"));
+        tx1.commit();
+        tx3.commit();
+    }
+
+    @Test
+    // LockObjectForModification is currently used only in map-jpa
+    @RequireProvider(value = MapStorageProvider.class, only = JpaMapStorageProviderFactory.PROVIDER_ID)
+    public void testLockObjectForModification() {
+        try {
+            CONFIG.spi(MapStorageSpi.NAME)
```

Comment:
This can / should be a no-op since `CONFIG` is generally parsed only upon factory initialization which could happen long before this test.

Reply:
Thanks for this, this was a very bad way to set the config. As discussed we will sort this out later (to unblock other work mainly in HotRod that depends on this) as part of this issue: [#16616](https://github.com/keycloak/keycloak/issues/16616). Until then we set `lockTimeout` to 1000 for all model tests.